### ISSUE TITLE
Run Functional Tests Hangs on 'InProgress' state (re-opened)

### ISSUE THREAD

**Issue by kalin-toshev** [Microsoft/vsts-tasks#5213]
_[re-opening the issue, since the original one ( #4684) is closed; pasting content below...]_

I hit the same issue (System.IO.FileLoadException/E_UNEXPECTED) while trying to hook up Functional Tests task, preceded by Deploy TestAgent in my VSTS release definition. I also spent some time to investigate the root cause and I think I found the problem. 

Basically, during startup, DTAExecutionHost.exe (via [Microsoft_VisualStudio_TestService_Utility]) installs an AssemblyResolver handler, which is supposed to resolve references to various assemblies, residing in different directories (under TestAgent2017 or Program Files (x86)) from the 'main' location (TestAgent). When the process goes through its initialization [Microsoft_VisualStudio_TestService_VstestAdapter!Microsoft.VisualStudio.TestService.VstestAdapter.SessionDetailStore..ctor] 'triggers' loading of Microsoft.VisualStudio.TestPlatform.ObjectModel.dll (which is under Program Files (x86)\...\TestWindow or under TestAgent2017), because of a referenced type (TestResult). This eventually hits the following stack - 

...
**0053b6f8 71220fc5 clr!AssemblyNative::Load+0x330**
0053b71c 712c9439 mscorlib_ni!System.Reflection.RuntimeAssembly.nLoad(System.Reflection.AssemblyName, System.String, System.Security.Policy.Evidence, System.Reflection.RuntimeAssembly, System.Threading.StackCrawlMark ByRef, IntPtr, Boolean, Boolean, Boolean)+0x25
0053b758 71257593 mscorlib_ni!System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(System.Reflection.AssemblyName, System.Security.Policy.Evidence, System.Reflection.RuntimeAssembly, System.Threading.StackCrawlMark ByRef, IntPtr, Boolean, Boolean, Boolean)+0x99
0053b780 7126d5bf mscorlib_ni!System.Reflection.RuntimeAssembly.InternalLoadFrom(System.String, System.Security.Policy.Evidence, Byte[], System.Configuration.Assemblies.AssemblyHashAlgorithm, Boolean, Boolean, System.Threading.StackCrawlMark ByRef)+0x67
0053b7a0 0560fb33 mscorlib_ni!System.Reflection.Assembly.LoadFrom(System.String)+0x23
0053b7c8 0560f62f Microsoft_VisualStudio_TestService_Utility!Microsoft.VisualStudio.TestService.Utility.TestPlatformAssemblyLoader.AssemblyResolve_LoadFrom(System.String, System.String, Boolean)+0xd3
0053b800 71223395 Microsoft_VisualStudio_TestService_Utility!Microsoft.VisualStudio.TestService.Utility.TestPlatformAssemblyLoader.CurrentDomain_AssemblyResolve(System.Object, System.ResolveEventArgs)+0x1b7
0053b824 7254eaf6 mscorlib_ni!System.AppDomain.OnAssemblyResolveEvent(System.Reflection.RuntimeAssembly, System.String)+0x95
...

Eventually (continuing 'stepping' the stack above), we will go through urlmon!CSecurityManager::MapUrlToZone (urlmon.dll is loaded) and from there, we land in urlmon!InternetCreateZoneManager. This is the function that **sometimes** fails with E_UNEXPECTED when the CUrlZoneManager fails to initialize (this propagates up the stack as the exception, seen above). The reason why it fails sometimes, is that **HKCU hive may not be loaded** at the time, when the call is made (and it's needed there!). Sometimes, when we are lucky, we may get the hive (profile) loaded via other tasks (or RDP session or another WinRm connection with the same account) at the time DTAExecutionHost.exe needs it (because of the Assembly.LoadFrom business above). But then in other times, the HKCU is not loaded and then we hit the error above. 
Because of the tricky conditions when this happens, I had to hook a remote debugger (IFEO/DTAExecutionHost.exe) using something like this - **c:\debuggers\cdb.exe -server tcp:port=4445 -noio** on the node (opened the port, etc.), running the tests; then make sure I am logged off (and all WinRm sessions using this account are closed). Then, when VSTS release workflow activates the DTAExecutionHost.exe, I connect from another box (cdb -remote tcp:server=<node ip>,port=4445) and wait for ~1 min (again to make sure that no WMI workers (wsmprovhost.exe - from previous WinRm connections) are around (keeping the hive loaded)). Then, **sxe ld:urlmon.dll; g** in the debugger. Once we stop in ntdll.dll, **bp urlmon!InternetCreateZoneManager** and **g** and then (when we stop), gu and check *eax* (register) -> E_UNEXPECTED (0x8000FFFF). 

Note that this can happen in any other 'context' where the HCKU hive is needed and we have a similar 'setup'. 

Currently, I patched things up by throwing in a 'wrapper' pair of tasks (basically a PS script over WinRm) - 'lock/unlock', During 'lock', I leave the (remote) session is 'disconnected' state (Disconnect-PSSession), which would keep the wsmprovhost.exe and the (hive) state on the node loaded. Once we are done, the 'unlock' step tries to reconnect to the (existing) session (Get-PSSession -> Connect-PSSession) and eventually removes it (cleanup). Note that unlock is set to always run in the release definition. After that ('patch'), I have not seen the issue anymore, but it would be nice if we have a real fix in the VSTS task(s) :-)

**Comment by nmaier26** [comment]
This has also been a major issue for us as well.  If there is anything more that I can add that hasn't already been provided let me know.  But wanted to comment to hopefully increase urgency.

**Comment by nigurr** [comment]
@kalin-toshev, @nmaier26  Can you select "Enable UI Testing" in Deploy Test agent task and try running the scenario and let us know is it still happening?

**Comment by nmaier26** [comment]
Selecting "Enable UI Testing" does appear to work around the issue, though not ideal as we aren't wanting to auto-login a user.

**Comment by kalin-toshev** [comment]
This was one of the first things that I tried, but it failed, because auto-logon could not be enabled.

**Comment by kalin-toshev** [comment]
For the workaround, I am using the attached PS script. I am calling the script 2 times - once in the beginning of the sequence and once at the end (with -unlock switch). Because I am disconnecting the PS (remote) session, the user's hive remains loaded and everything works. During unlock, I (try to) reconnect to the session and close it (which cleans things up). I think I pinpoint the problem above, right?  
[VstsAgentWinRmLock.zip](https://github.com/Microsoft/vsts-tasks/files/1278725/VstsAgentWinRmLock.zip)

**Comment by kalin-toshev** [comment]
Also, following the steps above, you can reliably reproduce the problem 100% (just make sure that no logon session is active for the selected user).

**Comment by nigurr** [comment]
@kalin-toshev We have tried creating new user and never logged in/no session exists with that user :)
But thanks for inputs. we will try little more

**Comment by kalin-toshev** [comment]
Let me provide a bit more information. 

In order to reproduce the problem, it's critical to allow DTAExecutionHost.exe to 'cache' the HKEY_CURRENT_USER (HANDLE) in the process (during startup) *before* the user hive is unloaded. In our case, this would happen during process startup here (from ETW trace) - 

[Root]
  ntdll.dll!_RtlUserThreadStart
  ntdll.dll!__RtlUserThreadStart
  kernel32.dll!BaseThreadInitThunk
  mscoree.dll!_CorExeMain_Exported
  mscoreei.dll!_CorExeMain
  mscoreei.dll!GetCorExeMainEntrypoint
  mscoreei.dll!CLRMetaHostPolicyImpl::GetRequestedRuntime
  mscoreei.dll!CLRMetaHostPolicyImpl::GetRequestedRuntimeHelper
  mscoreei.dll!RuntimeRequest::ComputeVersionString
  mscoreei.dll!RuntimeRequest::ComputeVersionStringThrowing
  |- mscoreei.dll!RuntimeRequest::ParseComponentConfig
  |- mscoreei.dll!ShimPolicies::ShimPolicies
  |    KernelBase.dll!RegOpenKeyExW
  |    KernelBase.dll!RegOpenKeyExInternalW
  |    KernelBase.dll!MapPredefinedHandleInternal
  |    KernelBase.dll!LocalOpenCurrentUser
  |    ntdll.dll!RtlOpenCurrentUser
  |    ntdll.dll!ZwOpenKey
  |    ntdll.dll!?
  |    ntdll.dll!?
...

Once the HKCU is cached in the process (pre-defined handle) while we still have a reference to the hive (held by the service - VisualStudioRemoteDeployer.exe), after the profile is unloaded (once the service terminates), *further references HKCU will fail with ERROR_KEY_DELETED* (3fa) (mapped from STATUS_HIVE_UNLOADED). This is not handled correctly by zone mapping code and eventually, we get E_UNEXPECTED (thrown as exception). In order to ensure that the above code will execute (caching the HKCU), we cannot simply stop at the initial breakpoint (in DTAExecutionHost.exe) - that's why I used -g option for the debugger in IFEO (more specifically, I have the following command line - **c:\debuggers_x86\cdb.exe -server tcp:port=4445 -noio -G -c "sxd av; sxe ld:clr.dll; gc; !soe -create System.IO.FileLoadException"**). If we do not do that, then we will give it enough time for the user hive to be unloaded before the HKCU handle is cached in the process, which will result in 'normal' ERROR_FILE_NOT_FOUND errors during registry query (that's why it 'works' sometimes), which is handled correctly by the URL zone mapping code (in urlmon). So the timing is really important here in order to see the problem. 

Attaching is the captured debugging session, showing the actual failure (by stepping through the code). As we can see, queries to HKCU fail with *3fa* (ERROR_KEY_DELETED). Looks like the remedy here would be to start the process with profile loaded (in order to keep a stable reference to it). 

[cdb.txt](https://github.com/Microsoft/vsts-tasks/files/1287254/cdb.txt)

**Comment by nigurr** [comment]
@kalin-toshev interesting. Can you try one simple thing in your local and see if it works (honestly, we tried without any sessions in our local lab, not able to repro)

In your local build agent can you change this [LoadUserProfile] to "true" - https://github.com/Microsoft/vsts-tasks/blob/master/Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1#L141

Really appreciating your help in this

**Comment by kalin-toshev** [comment]
Yes, I looked at that, but I do not think the suggestion above will work with the existing code. The reason is that LoadUserProfile is only 'taken into account' if we also have something in startInfo.UserName and this not case (when running as service - we run under the same account).  LoadUserProfile (LOGON_WITH_PROFILE) is only significant if startInfo.UserName is also set i.e. when Process.Start takes the *CreateProcessWithLogonW* path instead of just *CreateProcess*. 

I think is this case it would be easier if we somehow keep the service (VisualStudioRemoteDeployer.exe) around - this would make sure that the profile is still loaded while DTAExecutionHost.exe is running. Not sure if TestAgentConfiguration.ps1 would be the best place (it's convenient) - simply to WaitForExit() on the new process. Alternatively, if VisualStudioRemoteDeployer.exe expects the script to return, then the script has to pass the new process (or simply the ID) back and VisualStudioRemoteDeployer.exe can wait on it when it's appropriate. Once the DTAExecutionHost.exe is done, VisualStudioRemoteDeployer.exe can also go away. 

Btw, how do I try this 'locally'? Unfortunately, I do not have access to the subscription with our agents (there is some (shared) security state on them). Is there some way to 'push' new version of the task(s) somehow from the 'server side'? Also, if I had access to the agents, is there a local folder or something where I can drop the (local) version of the task so that I can try (just out of curiosity)? 

Again, in order to reproduce the problem reliably, we need a careful timing where the DTAExecutionHost.exe starts before the profile is unloaded (so that HKCU is cached) and then profile is unloaded before DTAExecutionHost.exe 'consumes' the HKCU (indirectly during the assembly resolution), which would fail with the error above (ERROR_KEY_DELETED). Unless you use the debugger, you may never be able to see this (even with the debugger, there is no guarantee, because it depends on when the service (VisualStudioRemoteDeployer.exe) will exits). That's why I think we need to keep the service around to ensure a stable user profile in DTAExecutionHost.exe.

**Comment by nigurr** [comment]
@kalin-toshev can you please try this? https://github.com/Microsoft/vsts-tasks/blob/master/Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1#L150
Can you explicitly add line 'Start-Sleep -Seconds 60' and try running the same (we tried different ways to repro in our labs, couldn't :-/ )

If our theory is correct, within that 60 seconds, the DTAExecutionHost will kick in the assembly resolver and loads the required dlls as the VS Remote deployer service is still active.
Once the dlls are loaded, the service can shut off and it should continue to work.

**Comment by kalin-toshev** [comment]
Yes, this would mitigate the problem as it would give a chance for the resolver to resolve the assemblies (at least in the beginning) while the hive is still loaded (as I said, I cannot try this unless it's deployed in VSTS, because I have no access to the agents), but it's not ideal, because if we (ever) need to resolve something after that (may never happen), we would still hit the issue. That's why I suggested to keep the service around while the execution host is running (if this is not possible, then sleep -sec 60 would still be better than nothing). 

The problem can be reproduced if you wait (in DTAExecutionHost) for some time (to make sure that the service is gone) in the beginning of the process before the resolver kicks in (I managed to do this with the debugger as described above).

**Comment by nigurr** [comment]
@kalin-toshev thanks for the info. I will try out that. We are moving towards "VsTest v2" task and deprecating the "Run Functional test" task. So I am just weighing the right fix given that service can't keep running [out of our control].

**Comment by nigurr** [comment]
Just an update, I tried adding 2min wait in the starting of DTAExecutionHost before Assembly Resolver kicks in and made sure that no user sessions are active, and still working in our labs. Never ran into UNEXPECTED error :(

**Comment by kalin-toshev** [comment]
Something (process) must be keeping the hive from unloading; note that it does not need to be an interactive process. Alternatively, you might have 'HKLM only option' (policy) turned on in your lab (for urlmon). 
You can also check when the hive is loaded/unloaded using xperf (ETW) - just make sure that you start a kernel session and logoff - this is what I used to prove what's happening - 

xperf -on Base+REG_HIVE -stackwalk RegHiveInit+RegHiveDestroy
<run the scenario, then stop the session...>
xperf -d registry.etl

In any case, adding the sleep should be able to address/mitigate the issue at this point.

**Comment by kalin-toshev** [comment]
We started hitting another issue while deploying the TestAgent :-( Attached is the log from VSTS. Note that it happens when the agent is not installed on the (remote) machine (first time run) and it fails after waiting for >1h. 

Is this a known issue? Looks like the deployer is trying to invoke PS script with invalid args. 

[DeployTestAgent.txt](https://github.com/Microsoft/vsts-tasks/files/1319311/DeployTestAgent.txt)

**Comment by nigurr** [comment]
@RanjanarMS Can you please look into this with priority?

**Comment by nmaier26** [comment]
I may have hit a similar issue that kalin-toshev.  Though in my case I believe it's due to the machine having all internet access blocked including VSTS (which I'm working to resolve).  Follow up question though, for the Visual Studio Test Agent Deployment task what domain and port access does it need to download and install the VS Test Agent?  I'll need to open these up on our firewall as well.  Thanks!

**Comment by nigurr** [comment]
@nmaier26 We require HTTPS port opened. However we do have offline test agent installation support. please refer https://blogs.msdn.microsoft.com/devops/2017/05/05/using-visual-studio-agent-deployment-task-on-machines-not-connected-to-the-internet/

**Comment by nmaier26** [comment]
Port 443, and what domain?

Offline installer is what we are trying to use currently.  We added it as a package in VSTS.  But it's not ideal for two reasons, 1) the size of the file is relatively large and it gets copied between multiple servers/machines making the task run long and 2) we would have to manually update the package whenever a new version comes out.

**Comment by nigurr** [comment]
@nmaier26 We suggest you following 
1. Install Test Agent in all those machines
2. Whitelist following urls for downloading the test execution engine https://testexecution.blob.core.windows.net/*
